tensorflow_addons/utils/python/tensor_util.py

```diff
@@ -0,0 +1,319 @@
+# Copyright 2015 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+"""Tensor utility functions."""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import numpy as np
+
+from tensorflow.python.framework import dtypes
+from tensorflow.python.framework import ops
+from tensorflow.python.framework import sparse_tensor
+from tensorflow.python.framework import tensor_shape
+from tensorflow.python.framework import tensor_util
+from tensorflow.python.ops import array_ops
+from tensorflow.python.ops import check_ops
+from tensorflow.python.ops import control_flow_ops
+from tensorflow.python.ops import math_ops
+from tensorflow.python.util.deprecation import deprecated
+
+
+__all__ = [
+    'assert_same_float_dtype',
+    'assert_scalar',
+    'assert_scalar_int',
+    'convert_to_tensor_or_sparse_tensor',
+    'is_tensor',
+    'reduce_sum_n',
+    'remove_squeezable_dimensions',
+    'with_shape',
+    'with_same_shape']
+
+
+# Temporary for backwards compatibility
+is_tensor = tensor_util.is_tensor
+assert_same_float_dtype = check_ops.assert_same_float_dtype
+assert_scalar = check_ops.assert_scalar
+
+convert_to_tensor_or_sparse_tensor = (
+    sparse_tensor.convert_to_tensor_or_sparse_tensor)
+
+
+def reduce_sum_n(tensors, name=None):
```

Comment:
Seeing as this exists as a general utility, can we decorate this with `@tf.function`? Not positive but I imagine there would likely be some hardware acceleration that can be done on this function in graph mode. 

To this point are there other places we should be utilizing `@tf.function` and autograph?

Reply:
By wrapping with tf.function, you are forcing it to be graph mode, which mean user loss debugability in eager mode. If user calls those python function by some code that is wrapped by tf.function, then this function will be picked by tf.function as well. So I am inclined to just not wrap them unless there is a strong reason do so.

Reply:
So I'm unclear about how much debug-ability will be lost, but this is certainly your call for the subpackage you'll maintain. We'll continue the discussion in #13 w.r.t the rest of addons.